### ISSUE TITLE
Bad memory behavior in ExceedMemoryAllocationCount

### ISSUE THREAD

**Issue by chrisforbes** [KhronosGroup/Vulkan-LoaderAndValidationLayers#1846]
Valgrind is very unhappy about this test:

```
[ RUN      ] VkLayerTest.ExceedMemoryAllocationCount
==69840== Invalid write of size 8
==69840==    at 0x4E4A11F: loader_phys_dev_ext_gpa (loader.c:3922)
==69840==    by 0x4E4EDDC: trampolineGetProcAddr (gpa_helper.h:175)
==69840==    by 0x4E4EDDC: vkGetInstanceProcAddr (trampoline.c:64)
==69840==    by 0x46DD84: VkLayerTest_ExceedMemoryAllocationCount_Test::TestBody() (layer_validation_tests.cpp:3048)
==69840==    by 0x50AC292: HandleSehExceptionsInMethodIfSupported<testing::Test, void> (gtest.cc:2078)
==69840==    by 0x50AC292: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2114)
==69840==    by 0x50A33F6: testing::Test::Run() (gtest.cc:2151)
==69840==    by 0x50A349D: testing::TestInfo::Run() (gtest.cc:2326)
==69840==    by 0x50A35A4: testing::TestCase::Run() (gtest.cc:2444)
==69840==    by 0x50A3857: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:4311)
==69840==    by 0x50A3B13: HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2078)
==69840==    by 0x50A3B13: HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2114)
==69840==    by 0x50A3B13: testing::UnitTest::Run() (gtest.cc:3925)
==69840==    by 0x441813: RUN_ALL_TESTS (gtest.h:2297)
==69840==    by 0x441813: main (layer_validation_tests.cpp:24849)
==69840==  Address 0xa06cec8 is 664 bytes inside an unallocated block of size 643,472 in arena "client"
==69840== 
==69840== Invalid read of size 8
==69840==    at 0x4E4A128: loader_phys_dev_ext_gpa (loader.c:3924)
==69840==    by 0x4E4EDDC: trampolineGetProcAddr (gpa_helper.h:175)
==69840==    by 0x4E4EDDC: vkGetInstanceProcAddr (trampoline.c:64)
==69840==    by 0x46DD84: VkLayerTest_ExceedMemoryAllocationCount_Test::TestBody() (layer_validation_tests.cpp:3048)
==69840==    by 0x50AC292: HandleSehExceptionsInMethodIfSupported<testing::Test, void> (gtest.cc:2078)
==69840==    by 0x50AC292: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2114)
==69840==    by 0x50A33F6: testing::Test::Run() (gtest.cc:2151)
==69840==    by 0x50A349D: testing::TestInfo::Run() (gtest.cc:2326)
==69840==    by 0x50A35A4: testing::TestCase::Run() (gtest.cc:2444)
==69840==    by 0x50A3857: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:4311)
==69840==    by 0x50A3B13: HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2078)
==69840==    by 0x50A3B13: HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2114)
==69840==    by 0x50A3B13: testing::UnitTest::Run() (gtest.cc:3925)
==69840==    by 0x441813: RUN_ALL_TESTS (gtest.h:2297)
==69840==    by 0x441813: main (layer_validation_tests.cpp:24849)
==69840==  Address 0xa06cec8 is 664 bytes inside an unallocated block of size 643,472 in arena "client"

... snip ...
```

**Comment by chrisforbes** [comment]
Possible that this was a driver issue. The machine I saw it on was running nvidia-367.55; after updating it to nvidia-381.22, valgrind is clean.

**Comment by mark-lunarg** [comment]
@chrisforbes, I think there **is** an issue -- this tests loads a layer containing an extension, and the loader has been relying on some extremely sketchy assumptions which break down in this corner-case.  It dies every time in some 32-bit environments and is indeterminately failing on even 64-bit builds.  Anyway, it is a loader issue and @lenny-lunarg is looking at it.

**Comment by chrisforbes** [comment]
Alright then, reopened :)

**Comment by lenny-lunarg** [comment]
To elaborate on what @mark-lunarg said, we found a loader issue in this test, where unknown physical device extensions rely on the compiler performing tail call optimizations, and the compiler wasn't performing that optimization, even when set to perform all optimizations. This was definitely the problem on Windows 32-bit, and it wouldn't surprise me if your problem is the same. But I'll look into the behavior on Linux once I finish fixing the Windows behavior.

**Comment by Tony-LunarG** [comment]
That looks like the same valgrind error I was getting during initial development on Linux, but it was accompanied by an access violation, which  @MarkY-LunarG fixed.  Apparently fixing the access violation did not fix the write/read that valgrind was upset with.

**Comment by lenny-lunarg** [comment]
@MarkY-LunarG fixed this today with commit d660c139f540f669a964a4466f709a7cd777333e. The problem was a memory allocation of the wrong size.
